osquery/events/windows/ntfs_event_publisher.h

```diff
@@ -0,0 +1,211 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#pragma once
+
+#include <atomic>
+#include <cstdint>
+#include <limits>
+#include <map>
+#include <memory>
+#include <unordered_set>
+
+#include <osquery/events.h>
+
+#include "osquery/events/windows/usn_journal_reader.h"
+
+namespace osquery {
+/// The subscription context contains the list of paths the subscriber is
+/// interested in
+struct NTFSEventSubscriptionContext : public SubscriptionContext {
+  /// The category that this event originated from.
+  std::string category;
+
+  /// Collection of paths that should only be included during write or
+  /// delete operations.
+  std::unordered_set<std::string> write_paths;
+
+  /// Collection of file reference numbers that should only be included
+  /// during write or delete operations.
+  std::unordered_set<USNFileReferenceNumber> write_frns;
+
+  /// Collection of paths that must always be included (even for reads).
+  std::unordered_set<std::string> access_paths;
+
+  /// Collection of file reference numbers that must always be included
+  /// (even for reads).
+  std::unordered_set<USNFileReferenceNumber> access_frns;
+
+ private:
+  friend class FileEventPublisher;
+};
+
+using NTFSEventSubscriptionContextRef =
+    std::shared_ptr<NTFSEventSubscriptionContext>;
+
+/// A single NTFS event record
+struct NTFSEventRecord final {
+  /// Event type
+  USNJournalEventRecord::Type type;
+
+  /// Path
+  std::string path;
+
+  /// Previous path (only valid for rename operations)
+  std::string old_path;
+
+  /// Record timestamp
+  std::time_t record_timestamp{0U};
+
+  /// Node attributes
+  DWORD attributes;
+
+  /// Update sequence number of the journal record
+  USN update_sequence_number{0U};
+
+  /// Ordinal for the file or directory referenced by this record
+  USNFileReferenceNumber node_ref_number;
+
+  /// Ordinal for the directory containing the file or directory referenced
+  USNFileReferenceNumber parent_ref_number;
+
+  /// Drive letter
+  char drive_letter{0U};
+
+  /// If true, this event is partial; it means that we could only get
+  /// the file or folder name inside path or old_path
+  bool partial{false};
+
+  NTFSEventRecord(const USNJournalEventRecord& rec)
+      : type(rec.type),
+        record_timestamp(rec.record_timestamp),
+        attributes(rec.attributes),
+        update_sequence_number(rec.update_sequence_number),
+        node_ref_number(rec.node_ref_number),
+        parent_ref_number(rec.parent_ref_number),
+        drive_letter(rec.drive_letter) {}
+};
+
+/// This structure is used to save volume handles and reference ids
+struct VolumeData final {
```

Comment:
> or making this data structure more RAII friendly, what do you think?

This sounds good to me! I can add a destructor that closes `volume_handle` and `root_folder_handle` appropriately.

Reply:
Ah, this might be a little annoying because of `volume_data_map` -- we need to keep the handles open beyond the function scope, since we insert each `VolumeData` (on success) into the map, which we only collect during publisher teardown.

Reply:
@woodruffw yeah scanning through some of this I was worried getting the `scope_guard`s working might be a bit tricky, and I wasn't sure if getting the RAII, with the VolumeData as a class, would work or not. If we can't get it sorted I think that's ok, but I'd like to run some handle leak testing on this just to be sure, as we've been caught with `HANDLE` leaks before and it was a bummer.